### PR TITLE
Add Builder API + more checkers

### LINKED ISSUE #15: Add builder API
This would allow for some similar logic to how gson and moshi builders are done.

```kotlin
// Simple case
moshi.interopWith(gson)

// Advanced case
moshi.interopBuilder(gson)
  .addMoshiType(TypeA::class.java) // with reified extensions
  .addGsonType(TypeB::class.java)
  .addMoshiFactory(TypeAFactory::class.java) // take a moshi adapter factory. If it returns non-null, it's a moshi type
  .addGsonFactory(TypeBFactory::class.java)
  .addClassChecker(...)
```

Then in a conventional setup where someone is maybe wiring a multibinding set of adapters, they can easily add the same types

Another idea is to take JsonAdapter/TypeAdapter methods directly and resolve their generics with reflection to indicate their type?

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #15 

This reworks the API a bit for more advanced usage (see tests for examples)

Builder adds a bit more fine-grained control:
* Logging option
* Default serializer control
* Public access to some built-ins for composing
* Better decision tree - checkers can offer "no opinion" to defer

### CODE REVIEW

File: src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+) : ClassChecker {
+  @Suppress("ReturnCount")
+  override fun serializerFor(rawType: Class<*>): Serializer? {
+    if (rawType.isEnum) {
+      @Suppress("UNCHECKED_CAST")
+      val constants: Array<out Enum<*>> = rawType.enumConstants as Array<out Enum<*>>
+      for (constant in constants) {
+        val field = rawType.getField(constant.name)
+        if (field.isAnnotationPresent(SerializedName::class.java)) {
+          logger?.invoke("🧠 Picking GSON for enum $rawType based on @SerializedName-annotated member $field.")
```
Comment (swolfand) [inline src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt:77]: Wild that this is valid

**Review by swolfand — APPROVED** [review]

File: src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+          return MOSHI
+        }
+      }
+      logger?.invoke("🧠 Defaulting to $defaultSerializer for enum $rawType.")
+      return defaultSerializer
+    }
+    return null
+  }
+
+  override fun toString(): String = "EnumClassChecker(defaultToMoshi=$defaultSerializer)"
```
Comment (adamsp) [inline src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt:90]:
```suggestion
  override fun toString(): String = "EnumClassChecker(defaultSerializer=$defaultSerializer)"
```

File: src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      return it
+    }
+
+    // Nothing else worked.
+    logger?.invoke("🧠 No decision for $rawType, defaulting to $defaultSerializer")
+    return defaultSerializer ?: error("No serializer found for $rawType")
+  }
+
+  override fun toString(): String {
+    return "StandardMoshiCheckers"
```
Comment (adamsp) [inline src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt:120]:
```suggestion
    return "StandardClassCheckers"
```

File: src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt (hunk 1 above)
Comment (ZacSweers) [reply src/main/kotlin/com/slack/moshi/interop/gson/ClassCheckers.kt:77]: 🧠

**Pull request merged**